# Getting Data
The topics of this week continues to be getting data, in this case using an API to access structured data. 

In this lab notebook you will gain experience reading data from and posting to an API. 


## Lab Setup

In [111]:
import requests
import json
import datetime
from io import StringIO
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl 
%matplotlib inline  

import otter
grader = otter.Notebook()

## API Getting Data

So far we have seen examples of getting data from an API.  These examples make use of GET requests from the API/server. 

Making a HTTP GET request can be done using several python libraries including: 

* httplib 
* urllib 
* requests 

We have been using the `requests` module.

Let's look at another example.

## Example: Google Books

Here we will examine using the Google Books API:  
https://developers.google.com/books/docs/overview


We will be using the "volumes" resource which does not require authentication.  
https://developers.google.com/books/docs/v1/getting_started#background-operations

Specifically, we will be using the query function to search by ISBN or book numbers. 
https://developers.google.com/books/docs/v1/using#PerformingSearch




In [112]:
# api-endpoint 
url = "https://www.googleapis.com/books/v1/volumes"
  
isbn = "isbn:0553386794"

# set the parameters to be sent to the API
params = {'q': isbn}

resp = requests.get(url, params)

Look at what the response is? 

How do we then extract the data?

In [113]:
resp

<Response [200]>

In [114]:
dat = resp.json()
#dat

# First, we can print it better! 
print(json.dumps(resp.json(), indent=2))

{
  "kind": "books#volumes",
  "totalItems": 2,
  "items": [
    {
      "kind": "books#volume",
      "id": "hXNvadj27ekC",
      "etag": "rvp+kdA8lxI",
      "selfLink": "https://www.googleapis.com/books/v1/volumes/hXNvadj27ekC",
      "volumeInfo": {
        "title": "A Game of Thrones (HBO Tie-in Edition)",
        "subtitle": "A Song of Ice and Fire: Book One",
        "authors": [
          "George R. R. Martin"
        ],
        "publisher": "Bantam",
        "publishedDate": "2011-03-22",
        "description": "NOW THE ACCLAIMED HBO SERIES GAME OF THRONES\u2014THE MASTERPIECE THAT BECAME A CULTURAL PHENOMENON Winter is coming. Such is the stern motto of House Stark, the northernmost of the fiefdoms that owe allegiance to King Robert Baratheon in far-off King\u2019s Landing. There Eddard Stark of Winterfell rules in Robert\u2019s name. There his family dwells in peace and comfort: his proud wife, Catelyn; his sons Robb, Brandon, and Rickon; his daughters Sansa and Arya; and hi

There is a lot of information here.  Explore the structure of the JSON information. 

In [115]:
dat.keys()

dict_keys(['kind', 'totalItems', 'items'])

In [116]:
dat['kind']

'books#volumes'

In [117]:
dat['totalItems']

2

In [118]:
dat['items']

[{'kind': 'books#volume',
  'id': 'hXNvadj27ekC',
  'etag': 'rvp+kdA8lxI',
  'selfLink': 'https://www.googleapis.com/books/v1/volumes/hXNvadj27ekC',
  'volumeInfo': {'title': 'A Game of Thrones (HBO Tie-in Edition)',
   'subtitle': 'A Song of Ice and Fire: Book One',
   'authors': ['George R. R. Martin'],
   'publisher': 'Bantam',
   'publishedDate': '2011-03-22',
   'description': 'NOW THE ACCLAIMED HBO SERIES GAME OF THRONES—THE MASTERPIECE THAT BECAME A CULTURAL PHENOMENON Winter is coming. Such is the stern motto of House Stark, the northernmost of the fiefdoms that owe allegiance to King Robert Baratheon in far-off King’s Landing. There Eddard Stark of Winterfell rules in Robert’s name. There his family dwells in peace and comfort: his proud wife, Catelyn; his sons Robb, Brandon, and Rickon; his daughters Sansa and Arya; and his bastard son, Jon Snow. Far to the north, behind the towering Wall, lie savage Wildings and worse—unnatural things relegated to myth during the centuries-l

###   
*HERE is where there is a lot of information* 


`dat['items']` returns a list of items, with a lot of information

In [119]:
type(dat['items'])

list

In [120]:
# We can look at the first item on the list 
dat['items'][0]

{'kind': 'books#volume',
 'id': 'hXNvadj27ekC',
 'etag': 'rvp+kdA8lxI',
 'selfLink': 'https://www.googleapis.com/books/v1/volumes/hXNvadj27ekC',
 'volumeInfo': {'title': 'A Game of Thrones (HBO Tie-in Edition)',
  'subtitle': 'A Song of Ice and Fire: Book One',
  'authors': ['George R. R. Martin'],
  'publisher': 'Bantam',
  'publishedDate': '2011-03-22',
  'description': 'NOW THE ACCLAIMED HBO SERIES GAME OF THRONES—THE MASTERPIECE THAT BECAME A CULTURAL PHENOMENON Winter is coming. Such is the stern motto of House Stark, the northernmost of the fiefdoms that owe allegiance to King Robert Baratheon in far-off King’s Landing. There Eddard Stark of Winterfell rules in Robert’s name. There his family dwells in peace and comfort: his proud wife, Catelyn; his sons Robb, Brandon, and Rickon; his daughters Sansa and Arya; and his bastard son, Jon Snow. Far to the north, behind the towering Wall, lie savage Wildings and worse—unnatural things relegated to myth during the centuries-long summer

In [121]:
'''We can investigate the keys where information is stored for each item'''
dat['items'][0].keys()

dict_keys(['kind', 'id', 'etag', 'selfLink', 'volumeInfo', 'saleInfo', 'accessInfo', 'searchInfo'])

In [122]:
# You can start building pretty long lines of code to access information deep 
#  in the structure. 
# Print out the ISBN_10 number for the book 
dat['items'][0]['volumeInfo']['industryIdentifiers'][0]['identifier']

'9780553386790'

## Exercise 1 

Which of the Game of Thrones books is longest?

Get information about each book and print out the title and number of pages.  Then, report the book title and number of pages for the book with that is the longest.  

*Note, the API may return multiple entries for each isbn.  You may use the first entry for information.  If the information is missing a page number it is likely an audiobook, and you should then use the next entry for information.  If no entry has the title and page number information return the title as "no title" and the number of pages as '-1'.*

In [123]:
''' Following is the isbn codes for Game of Thrones books. '''

isbns = ['0553386794', '0345535421', '9780345543981', '0553390570', '1101886048']

In [124]:
'''
Iterate for each isbns to finds titles and pages for each item. 
Collect this information in a list. 
Look to use "volumeInfo" to gather the information needed.
Print the title + the number of pages in the loop. 

Outside the loop:
- Convert the list to a DataFrame, ex1df, column names 'Title' and 'NumPages' 
- Report the longest book in longest8dbaec7bed057df01f5566cfc6499da8BookTitle and longestBookNumPages.
'''

ex1list = [] 

for i in isbns: 
    params = {'q': 'isbn:' + i}
    # passin the params to get 
    resp = requests.get(url, params)
    # loadind the data of the books
    dat = resp.json()
    # checking if the items are present or not 
    if 'items' in dat:
        # getting the voulmeinfo of the book 
        book_info = dat['items'][0]['volumeInfo']
        # getting the title of the book if no title is found returning no title 
        title = book_info.get('title', 'No title')
        # getting the no of pages of the book if pagecount is not found returning -1
        pages = book_info.get('pageCount', -1)
        # Adding the title and NoOfPages to the list
        ex1list.append({'Title': title, 'NumPages': pages})
    print(title + " has " + str(pages) + " pages.")
    
# creating the datframe by using the list 
ex1df = pd.DataFrame(ex1list, columns=['Title', 'NumPages'])

# getting the longest book from the dataframe
longest_book = ex1df[ex1df['NumPages'] == ex1df['NumPages'].max()]

#getting the title of the longest book title 
longestBookTitle = longest_book['Title'].values[0]

#getting the number of pages from the longest book 
longestBookNumPages =  longest_book['NumPages'].values[0]

A Game of Thrones (HBO Tie-in Edition) has 722 pages.
A Clash of Kings (HBO Tie-in Edition) has 1042 pages.
A Storm of Swords (HBO Tie-in Edition): A Song of Ice and Fire: Book Three has 1218 pages.
A Feast for Crows (HBO Tie-in Edition): A Song of Ice and Fire: Book Four has 786 pages.
A Dance with Dragons (HBO Tie-in Edition): A Song of Ice and Fire: Book Five has 1058 pages.


In [125]:
grader.check("q1")

q1 results: All test cases passed!

## Example: Government API - Iceland

Many cities or countries have begun making data available for developers and researchers. 
Iceland has created a single [API](http://docs.apis.is/) that has many endpoints including weather data, concerts, bus, earthquakes, bicycle counters, etc. 

Try out a few different endpoints and gather some data.  *Note, some of the endpoints are not available at this time, you may get a 404 or 500 error*

**Earthquake Information**

In [126]:
resp = requests.get('https://apis.is/earthquake/is')

In [127]:
resp

<Response [200]>

In [128]:
#resp.json()
# UNCOMMENT TO EXPLORE
# CHECK!!! PUT COMMENT BACK IN BEFORE SUBMISSION

Under the `results` key there is a list of items with the earthquake information. 

Let's try to get this list in a DataFrame.

In [129]:
# The JSON was already stored in "resp" and we only want the list of results 
#  under the key "results".  Therefore, we can take this information and 
#  re-serialize the information and let pandas read_json parse in the data 
eq = pd.read_json(json.dumps(resp.json()['results']))
eq.head()

timestamp  latitude  longitude  depth  size  quality  \
0 2017-10-13 12:07:24+00:00    63.976    -21.949    1.1   0.6    58.73   
1 2017-10-13 09:50:50+00:00    65.124    -16.288    7.2   0.9    78.51   
2 2017-10-13 09:41:09+00:00    63.945    -21.143    7.4   0.2    33.12   
3 2017-10-13 09:37:45+00:00    65.114    -16.300    6.3   1.2    90.01   
4 2017-10-13 09:37:21+00:00    65.113    -16.301    5.9   1.4    90.01   

             humanReadableLocation  
0          6,1 km SV af Helgafelli  
1  6,1 km NA af Herðubreiðartöglum  
2         6,5 km SSA af Hveragerði  
3  5,0 km NA af Herðubreiðartöglum  
4  4,9 km NA af Herðubreiðartöglum

If you get a "Value Error: protocol not known", then try wrapping the `json.dumps(resp.json()['results'])` in a method to write to a string using `StringIO` 

```python
from io import StringIO
eq = pd.read_json(StringIO(json.dumps(resp.json()['results'])))
```

**Football**

We can get information on football events. 



In [130]:
resp = requests.get('https://apis.is/sports/football')
resp

<Response [200]>

In [131]:
resp.json()

{'results': []}

Looks like no information listed for upcoming football matches. 

**Currency** 

In [132]:
params = {'source' : 'lb'}
resp = requests.get('https://apis.is/currency', params)
resp

<Response [200]>

In [133]:
resp.json()

{'results': [{'shortName': 'ISK',
   'longName': 'Íslensk króna',
   'value': 1,
   'askValue': 1,
   'bidValue': 1,
   'changeCur': 0,
   'changePer': '0.00'},
  {'shortName': 'USD',
   'longName': 'Bandarískur dalur',
   'value': 138.735,
   'askValue': 139.22,
   'bidValue': 138.25,
   'changeCur': -0.114794,
   'changePer': 0},
  {'shortName': 'GBP',
   'longName': 'Sterlingspund',
   'value': 169.205,
   'askValue': 169.8,
   'bidValue': 168.61,
   'changeCur': 0.082518,
   'changePer': '0.00'},
  {'shortName': 'EUR',
   'longName': 'Evra',
   'value': 146.3,
   'askValue': 146.81,
   'bidValue': 145.79,
   'changeCur': 0,
   'changePer': '0.00'},
  {'shortName': 'CAD',
   'longName': 'Kanadískur dalur',
   'value': 101.885,
   'askValue': 102.24,
   'bidValue': 101.53,
   'changeCur': 0.097905,
   'changePer': '0.00'},
  {'shortName': 'DKK',
   'longName': 'Dönsk króna',
   'value': 19.6125,
   'askValue': 19.681,
   'bidValue': 19.544,
   'changeCur': -0.010161,
   'changePer': 

**Weather Observations**

In [134]:
params = {'stations': '1'}
resp = requests.get('https://apis.is/weather/observations/en', params)
resp


<Response [200]>

In [135]:
resp.json()

{'results': [{'name': 'Reykjavík',
   'time': '2023-10-16 22:00:00',
   'err': '',
   'link': 'http://en.vedur.is/weather/observations/areas/reykjavik/#group=100&station=1',
   'F': '3',
   'FX': '3',
   'FG': '5',
   'D': 'ENE',
   'T': '3.3',
   'W': '',
   'V': '',
   'N': '',
   'P': '1008',
   'RH': '85',
   'SNC': '',
   'SND': '',
   'SED': '',
   'RTE': '',
   'TD': '1.0',
   'R': '0.0',
   'id': '1',
   'valid': '1'}]}

A description on what the codings stand for is available in the API documentation:  
http://docs.apis.is/#endpoint-weather

## Exercise 2 

Look at forecasts for the Reykjavík weather station (station : 1). Load this data into a DataFrame. Report the mean temperature and mean wind speed for the next 24 forecasts. 

In [136]:
# Use the Iceland API to collect weather forecasts for the Reykjavik station. 
# Store the next 24 weather forecasts in a DataFrame, wf 
# The column names should be the English descriptors for the variables collected
#   replaces spaces with _ and remove the parenthesis and anything after it in 
#   the description. 
#   For example: "Date", "Wind_speed", etc.
# Report the mean temperature and mean wind speed for the next 24 forecasts. 

params = {'stations' : '1'}
resp = requests.get('https://apis.is/weather/forecasts/en', params)
dat =  resp.json()

# Using the information above given 
# The JSON was already stored in "resp" and we only want the list of results 
#  under the key "results".  Therefore, we can take this information and 
#  re-serialize the information and let pandas read_json parse in the data 
    
wf = pd.read_json((json.dumps(resp.json()['results'][0]['forecast'])))

wf.columns = ['Date', 'Wind_speed', 'Wind_direction', 'Air_temperature', 'Weather_description', 'Cloud_cover', 'Dew_limit', 'Cumulative_precipitation']

wf = wf.head(24)
meanTemp = wf['Air_temperature'].astype(float).mean()
meanWS = wf['Wind_speed'].astype(float).mean() 

print("Mean Temperature (deg C): ", meanTemp)
print("Mean Wind Speed (m/s)   : ", meanWS)

Mean Temperature (deg C):  7.458333333333333
Mean Wind Speed (m/s)   :  8.291666666666666


In [137]:
wf.head()

Date  Wind_speed Wind_direction  Air_temperature  \
0  2023-10-17 00:00:00           4            ESE                4   
1  2023-10-17 01:00:00           4            ESE                4   
2  2023-10-17 02:00:00           5            ESE                4   
3  2023-10-17 03:00:00           5            ESE                4   
4  2023-10-17 04:00:00           5            ESE                4   

  Weather_description  Cloud_cover  Dew_limit  Cumulative_precipitation  
0            Overcast          100          1                       0.0  
1            Overcast          100          1                       0.0  
2            Overcast          100          1                       0.0  
3          Light rain          100          1                       0.1  
4                Rain          100          2                       0.5

In [138]:
grader.check("q2")

q2 results: All test cases passed!

## Example: iTunes Content 

Apple has a simple [API](https://affiliate.itunes.apple.com/resources/documentation/itunes-store-web-service-search-api/) for looking up iTunes content.

In [139]:
# api-endpoint
url = 'https://itunes.apple.com/search'

# For example let's search for lord of the rings ebooks 
params = {'term': 'lord+of+the+rings', 'entity': 'ebook', 
         'limit': 3}

resp = requests.get(url, params)

In [140]:
resp

<Response [200]>

In [141]:
resp.json()

{'resultCount': 1,
 'results': [{'trackViewUrl': 'https://books.apple.com/us/book/lord-of-the-rings/id739542595?uo=4',
   'trackCensoredName': 'Lord of the Rings',
   'fileSizeBytes': 501517,
   'formattedPrice': '$24.99',
   'artworkUrl60': 'https://is1-ssl.mzstatic.com/image/thumb/Publication/v4/5a/21/24/5a2124b1-b759-fd4c-3d1e-aa4c0c1e6641/9780813138015.jpg/60x60bb.jpg',
   'artworkUrl100': 'https://is1-ssl.mzstatic.com/image/thumb/Publication/v4/5a/21/24/5a2124b1-b759-fd4c-3d1e-aa4c0c1e6641/9780813138015.jpg/100x100bb.jpg',
   'artistViewUrl': 'https://books.apple.com/us/artist/jane-chance/482333908?uo=4',
   'artistIds': [482333908],
   'artistId': 482333908,
   'artistName': 'Jane Chance',
   'genres': ['Literary Criticism', 'Books', 'Fiction & Literature'],
   'price': 24.99,
   'description': '" With New Line Cinema\'s production of The Lord of the Rings film trilogy, the popularity of the works of J.R.R. Tolkien is unparalleled. Tolkien\'s books continue to be bestsellers deca

## Exercise 3

Search for the 50 "The Expanse" e-books (search may return fewer). Create a data frame from the responses containing the `trackName`, `track ID`, `price`, and `averageUserRating`. Sort the results from highest to lowest price.

If any of the information you are meant to collect (`trackName`, `track ID`, `price`, and `averageUserRating`) is missing, replace with `NaN`

In [142]:
url = 'https://itunes.apple.com/search'

# """ For example let's search for "The Expanse" ebooks """

params = {'term': 'expanse', 'entity': 'ebook' }
resp = requests.get(url, params) 

#resp.json()

In [143]:
obj = json.loads(resp.text)
# Alternatively 
obj2 = resp.json()

Try using at least two approaches to create the DataFrame, e.g., 

* *Method 1* - Keep track of rows in a list, convert nested lists to DataFrame.  Note, do not create an empty DataFrame and append entries in an iterator (this is not scalable)  
https://stackoverflow.com/questions/13784192/creating-an-empty-pandas-dataframe-and-then-filling-it/41529411#41529411
* *Method 2* - Use pandas `read_json` function to convert JSON to pandas object
* *Method 3* - Use `json_normalize` function to read in JSON to a flat table. 
The `json_normalize` function normalizes a semi-structured JSON data object into a flat table.   
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.json_normalize.html

<!-- BEGIN QUESTION -->



In [144]:
# State which method you are using: Method 1 
#  Method 1

# Creating a function to get the information if not found filling with NaN value
def extracting_info(result):
    track_name = result.get('trackName', 'NaN')
    track_id = result.get('trackId', 'NaN')
    price = result.get('price', 'NaN')
    avg_user_rating = result.get('averageUserRating', 'NaN')
    return [track_name, track_id, price, avg_user_rating]

# Getting the results 
results = obj2.get('results', [])

# Creating a list from the results
dat = [extracting_info(result) for result in results]

#using the list to create a dataframe 
q3df1 = pd.DataFrame(dat, columns=['trackName', 'trackID', 'price', 'averageUserRating'])

# Converting the price column into numeric values any non numneric values will be handled by errors
#https://pandas.pydata.org/docs/reference/api/pandas.to_numeric.html
q3df1['price'] = pd.to_numeric(q3df1['price'], errors='coerce')

# Sorting the dataframe by price column in descending order highest to lowest 
q3df1 = q3df1.sort_values(by='price', ascending=False)

print(q3df1.shape)
q3df1.head()

(50, 4)


trackName     trackID  price averageUserRating
2    Caliban's War   469995594  11.99               4.5
3   Abaddon's Gate   576438197  11.99               4.5
4      Cibola Burn   721822343  11.99               4.5
5    Nemesis Games   926360337  11.99               4.5
6  Babylon's Ashes  1063592195  11.99               4.5

<!-- END QUESTION -->

<!-- BEGIN QUESTION -->



In [145]:
# State which method you are using: Method 2  
#  Method 2 

# Using read_json to create dataframe 
q3df2 = pd.read_json(json.dumps(obj2['results']))

# Sorting the dataframe by price column in descending order highest to lowest 
q3df2 = q3df2.sort_values(by='price', ascending=False)

# Converting the price column into numeric values any non numneric values will be handled by errors
#https://pandas.pydata.org/docs/reference/api/pandas.to_numeric.html
q3df2['price'] = pd.to_numeric(q3df2['price'], errors='coerce')

#creating dataframe with required columns 
q3df2 = q3df2[['trackName', 'trackId', 'price', 'averageUserRating']]

print(q3df2.shape)
q3df2.head()

(50, 4)


trackName     trackId  price  averageUserRating
2    Caliban's War   469995594  11.99                4.5
3   Abaddon's Gate   576438197  11.99                4.5
4      Cibola Burn   721822343  11.99                4.5
5    Nemesis Games   926360337  11.99                4.5
6  Babylon's Ashes  1063592195  11.99                4.5

In [146]:
# State which method you are using: Method 3  
#  Method 3 

# Using read_json to create dataframe 
q4df4 = pd.json_normalize(obj2['results'])

# Converting the price column into numeric values any non numneric values will be handled by errors
q4df4['price'] = pd.to_numeric(q4df4['price'], errors='coerce')

# Sort the DataFrame by price in descending order
q4df4 = q4df4.sort_values(by='price', ascending=False)
#creating dataframe with required columns 
q4df4 = q4df4[['trackName', 'trackId', 'price', 'averageUserRating']]

q4df4.head()

trackName     trackId  price  averageUserRating
2    Caliban's War   469995594  11.99                4.5
3   Abaddon's Gate   576438197  11.99                4.5
4      Cibola Burn   721822343  11.99                4.5
5    Nemesis Games   926360337  11.99                4.5
6  Babylon's Ashes  1063592195  11.99                4.5

<!-- END QUESTION -->

## Example: TV Shows 

Here we can use an API on tv show information:  
http://api.tvmaze.com/

In [147]:
# We can find the tvmaze id for a show based on the IMDB id. 
id_got = 'tt3032476'
resp = requests.get('http://api.tvmaze.com/lookup/shows?imdb=tt3032476')

In [148]:
resp.json()

{'id': 618,
 'url': 'https://www.tvmaze.com/shows/618/better-call-saul',
 'name': 'Better Call Saul',
 'type': 'Scripted',
 'language': 'English',
 'genres': ['Drama', 'Crime', 'Legal'],
 'status': 'Ended',
 'runtime': 60,
 'averageRuntime': 64,
 'premiered': '2015-02-08',
 'ended': '2022-08-15',
 'officialSite': 'https://www.amc.com/shows/better-call-saul--1002228',
 'schedule': {'time': '21:00', 'days': ['Monday']},
 'rating': {'average': 8.6},
 'weight': 97,
 'network': {'id': 20,
  'name': 'AMC',
  'country': {'name': 'United States',
   'code': 'US',
   'timezone': 'America/New_York'},
  'officialSite': None},
 'webChannel': None,
 'dvdCountry': None,
 'externals': {'tvrage': 37780, 'thetvdb': 273181, 'imdb': 'tt3032476'},
 'image': {'medium': 'https://static.tvmaze.com/uploads/images/medium_portrait/399/998743.jpg',
  'original': 'https://static.tvmaze.com/uploads/images/original_untouched/399/998743.jpg'},
 'summary': '<p><b>Better Call Saul</b> is the prequel to the award-winni

We now know the TVmaze ID for Better Call Saul is **618**. 

## Exercise 4 

Calculate and report the min, mean, and max running time for Better Call Saul episodes by season in a DataFrame. 
Rows are indexed by season number (e.g., 1, 2, 3, ...) and Columns are "Min", "Mean", and "Max". 

Suggestion: 

* Create DataFrame of episodes information 
* Consider using the endpoint - http://www.tvmaze.com/api#show-episode-list
* Then use .groupby function to group by season
* Construct final DataFrame "bcs" to contain the requested information

In [149]:
# Create a DataFrame "bcs" with row index of season number and 
#  columns of "Min", "Mean" and "Max" running time of the episodes for that season.

# Getting the show info 
show_info = resp.json()
# Getting the show ID for Better Call Saul
show_id = show_info['id']
# Using the show ID to get the episodes 
resp = requests.get(f'http://api.tvmaze.com/shows/{show_id}/episodes')
# storing the info of episodes 
episodes_info = resp.json()
# Creating the dataframe for episodes
bcs = pd.DataFrame(episodes_info)
# Grouping the episodes by season 
bcs = bcs.groupby('season')
# calculating min, mean, and max running time using aggregated function  
bcs = bcs['runtime'].agg(['min', 'mean', 'max'])
# Renaming the columns
bcs.columns = ['Min', 'Mean', 'Max']
# Display the DataFrame
bcs

Min       Mean  Max
season                     
1        60  60.000000   60
2        60  60.000000   60
3        70  71.300000   77
4        60  60.000000   60
5        60  72.300000   85
6        60  60.923077   72

In [99]:
grader.check("q4")

q4 results: All test cases passed!

## Submission

Make sure you have run all cells in your notebook in order before running the cell below, so that all images/graphs appear in the output. The cell below will generate a zip file for you to submit. **Please save before exporting!**

In [150]:
# Save your notebook first, then run this cell to export your submission.
grader.export(pdf=False, run_tests=True)

Running your submission against local test cases...



Your submission received the following results when run against available test cases:

    q1 results: All test cases passed!

    q2 results: All test cases passed!

    q4 results: All test cases passed!
